packages/core/src/routes/applications/application.ts

```diff
@@ -166,29 +168,42 @@ export default function applicationRoutes<T extends AuthedRouter>(
         );
       }
 
-      const application = await insertApplication({
-        id: generateStandardId(),
-        secret: generateStandardSecret(),
-        oidcClientMetadata: buildOidcClientMetadata(oidcClientMetadata),
-        ...conditional(
-          rest.type === ApplicationType.Protected &&
-            protectedAppMetadata &&
-            (await protectedApps.checkAndBuildProtectedAppData(protectedAppMetadata))
-        ),
-        ...rest,
-      });
+      try {
+        const application = await insertApplication({
+          id: generateStandardId(),
+          secret: generateStandardSecret(),
+          oidcClientMetadata: buildOidcClientMetadata(oidcClientMetadata),
+          ...conditional(
+            rest.type === ApplicationType.Protected &&
+              protectedAppMetadata &&
+              (await protectedApps.buildProtectedAppData(protectedAppMetadata))
+          ),
+          ...rest,
+        });
 
-      if (application.type === ApplicationType.Protected) {
-        try {
-          await protectedApps.syncAppConfigsToRemote(application.id);
-        } catch (error: unknown) {
-          // Delete the application if failed to sync to remote
-          await deleteApplicationById(application.id);
-          throw error;
+        if (application.type === ApplicationType.Protected) {
+          try {
+            await protectedApps.syncAppConfigsToRemote(application.id);
+          } catch (error: unknown) {
+            // Delete the application if failed to sync to remote
+            await deleteApplicationById(application.id);
+            throw error;
+          }
         }
-      }
 
-      ctx.body = application;
+        ctx.body = application;
+      } catch (error: unknown) {
+        if (
+          error instanceof UniqueIntegrityConstraintViolationError &&
+          error.constraint === protectedAppMetadataHostConstraint
+        ) {
+          throw new RequestError({
+            code: 'application.protected_application_subdomain_exists',
+            status: 422,
+          });
+        }
```

Comment:
can move this to the koaErrorHandler level?

Reply:
updated